man/ffi.3
.Dd February 15, 2008
.Dt FFI 3
.Sh NAME
.Nm FFI
.Nd Foreign Function Interface
.Sh LIBRARY
libffi, -lffi
.Sh SYNOPSIS
.In ffi.h
.Ft ffi_status
.Fo ffi_prep_cif
.Fa "ffi_cif *cif"
.Fa "ffi_abi abi"
.Fa "unsigned int nargs"
.Fa "ffi_type *rtype"
.Fa "ffi_type **atypes"
.Fc
.Ft void
.Fo ffi_prep_cif_var
.Fa "ffi_cif *cif"
.Fa "ffi_abi abi"
.Fa "unsigned int nfixedargs"
.Fa "unsigned int ntotalargs"
.Fa "ffi_type *rtype"
.Fa "ffi_type **atypes"
.Fc
.Ft void
.Fo ffi_call
.Fa "ffi_cif *cif"
.Fa "void (*fn)(void)"
.Fa "void *rvalue"
.Fa "void **avalue"
.Fc
.Sh DESCRIPTION
The foreign function interface provides a mechanism by which a function can
generate a call to another function at runtime without requiring knowledge of
the called function's interface at compile time.

The full manual could be found in
.Nm info libffi\fR.
.Sh SEE ALSO
info libffi,
.Xr ffi_prep_cif 3 ,
.Xr ffi_prep_cif_var 3 ,
.Xr ffi_call 3


ChangeLog
2014-03-10  Gervase Markham <gerv@gerv.net>

	* msvcc.sh: Upgrade license to MPL 2.

2014-03-01  Anthony Green  <green@moxielogic.com>

	* Makefile.am (EXTRA_DIST): Replace old scripts with
	generate-darwin-source-and-headers.py.
	* Makefile.in: Rebuilt.

2014-02-28  Anthony Green  <green@moxielogic.com>

	* Makefile.am (AM_CFLAGS): Reintroduce missing -DFFI_DEBUG for
	--enable-debug builds.
	* Makefile.in: Rebuilt.

2014-02-28  Makoto Kato  <m_kato@ga2.so-net.ne.jp>

	* src/closures.c: Fix build failure when using clang for Android.

2014-02-28  Marcin Wojdyr  <wojdyr@gmail.com>

	* libffi.pc.in (toolexeclibdir): use -L${toolexeclibdir} instead
	of -L${libdir}.

2014-02-28  Paulo Pizarro  <paulo.pizarro@gmail.com>

	* src/bfin/sysv.S: Calling functions in shared libraries requires
	considering the GOT.

2014-02-28  Josh Triplett  <josh@joshtriplett.org>

	* src/x86/ffi64.c (classify_argument): Handle case where
	FFI_TYPE_LONGDOUBLE == FFI_TYPE_DOUBLE.

2014-02-28  Anthony Green  <green@moxielogic.com>

	* ltmain.sh: Generate with libtool-2.4.2.418.
	* m4/libtool.m4, m4/ltoptions.m4, m4/ltversion.m4: Ditto.
	* configure: Rebuilt.

2014-02-28  Dominik Vogt  <vogt@linux.vnet.ibm.com>

	* configure.ac (AC_ARG_ENABLE struct): Fix typo in help
	message.
	(AC_ARG_ENABLE raw_api): Ditto.
	* configure, fficonfig.h.in: Rebuilt.

2014-02-28  Will Newton  <will.newton@linaro.org>

	* src/arm/sysv.S: Initialize IP register with FP.

2014-02-28  Yufeng Zhang  <yufeng.zhang@arm.com>

	* src/aarch64/sysv.S (ffi_closure_SYSV): Use x29 as the
	main CFA reg; update cfi_rel_offset.

2014-02-09  Heiher  <r@hev.cc>

	* src/mips/n32.S: Fix call floating point va function.

2014-01-21  Zachary Waldowski  <zach@waldowski.me>

	* src/aarch64/ffi.c: Fix missing semicolons on assertions under
	debug mode.

2013-12-30  Zachary Waldowski  <zach@waldowski.me>

	* .gitignore: Exclude darwin_* generated source and build_* trees.
	* src/aarch64/ffi.c, src/arm/ffi.c, src/x86/ffi.c: Inhibit Clang
	previous prototype warnings.
	* src/arm/ffi.c: Prevent NULL dereference, fix short type warning
	* src/dlmalloc.c: Fix warnings from set_segment_flags return type,
	and the native use of size_t for malloc on platforms
	* src/arm/sysv.S: Use unified syntax. Clang clean-ups for
	ARM_FUNC_START.
	* generate-osx-source-and-headers.py: Remove.
	* build-ios.sh: Remove.
	* libffi.xcodeproj/project.pbxproj: Rebuild targets. Include
	x86_64+aarch64 pieces in library.  Export headers properly.
	* src/x86/ffi64.c: More Clang warning clean-ups.
	* src/closures.c (open_temp_exec_file_dir): Use size_t.
	* src/prep_cif.c (ffi_prep_cif_core): Cast ALIGN result.
	* src/aarch64/sysv.S: Use CNAME for global symbols.  Only use
	.size for ELF targets.
	* src/aarch64/ffi.c: Clean up for double == long double.  Clean up
	from Clang warnings.  Use Clang cache invalidation builtin.  Use
	size_t in place of unsigned in many places.  Accommodate for
	differences in Apple AArch64 ABI.

2013-12-02  Daniel Rodríguez Troitiño  <drodrigueztroitino@yahoo.es>

	* generate-darwin-source-and-headers.py: Clean up, modernize,
	merged version of previous scripts.

2013-11-21  Anthony Green  <green@moxielogic.com>

	* configure, Makefile.in, include/Makefile.in, include/ffi.h.in,
	man/Makefile.in, testsuite/Makefile.in, fficonfig.h.in: Rebuilt.

2013-11-21  Alan Modra  <amodra@gmail.com>

	* Makefile.am (EXTRA_DIST): Add new src/powerpc files.
	(nodist_libffi_la_SOURCES <POWERPC, POWERPC_FREEBSD>): Likewise.
	* configure.ac (HAVE_LONG_DOUBLE_VARIANT): Define for powerpc.
	* include/ffi.h.in (ffi_prep_types): Declare.
	* src/prep_cif.c (ffi_prep_cif_core): Call ffi_prep_types.
	* src/types.c (FFI_NONCONST_TYPEDEF): Define and use for
	HAVE_LONG_DOUBLE_VARIANT.
	* src/powerpc/ffi_powerpc.h: New file.
	* src/powerpc/ffi.c: Split into..
	* src/powerpc/ffi_sysv.c: ..new file, and..
	* src/powerpc/ffi_linux64.c: ..new file, rewriting parts.
	* src/powerpc/ffitarget.h (enum ffi_abi): Rewrite powerpc ABI
	selection as bits controlling features.
	* src/powerpc/linux64.S: For consistency, use POWERPC64 rather
	than __powerpc64__.
	* src/powerpc/linux64_closure.S: Likewise.
	* src/powerpc/ppc_closure.S: Likewise.  Move .note.FNU-stack
	inside guard.
	* src/powerpc/sysv.S: Likewise.
	* configure: Regenerate.
	* fficonfig.h.in: Regenerate.
	* Makefile.in: Regenerate.

2013-11-20  Alan Modra  <amodra@gmail.com>

	* src/powerpc/ffi.c (ffi_prep_cif_machdep_core): Use
	NUM_FPR_ARG_REGISTERS64 and NUM_GPR_ARG_REGISTERS64 not their
	32-bit versions for 64-bit code.
	* src/powerpc/linux64_closure.S: Don't use the return value area
	as a parameter save area on ELFv2.

2013-11-18  Iain Sandoe  <iain@codesourcery.com>

	* src/powerpc/darwin.S (EH): Correct use of pcrel FDE encoding.
	* src/powerpc/darwin_closure.S (EH): Likewise. Modernise picbase
	labels.

2013-11-18  Anthony Green  <green@moxielogic.com>

	* src/arm/ffi.c (ffi_call): Hoist declaration of temp to top of
	function.
	* src/arm/ffi.c (ffi_closure_inner): Moderize function declaration
	to appease compiler.
	Thanks for Gregory P. Smith <greg@krypto.org>.

2013-11-18  Anthony Green  <green@moxielogic.com>

	* README (tested): Mention PowerPC ELFv2.

2013-11-16  Alan Modra  <amodra@gmail.com>

	* src/powerpc/ppc_closure.S: Move errant #endif to where it belongs.
	Don't bl .Luint128.

2013-11-16  Alan Modra  <amodra@gmail.com>

	* src/powerpc/ffi.c (ffi_prep_cif_machdep_core): Use #if _CALL_ELF
	test to select parameter save sizing for ELFv2 vs. ELFv1.
	* src/powerpc/ffitarget.h (FFI_V2_TYPE_FLOAT_HOMOG,
	FFI_V2_TYPE_DOUBLE_HOMOG, FFI_V2_TYPE_SMALL_STRUCT): Define.
	(FFI_TRAMPOLINE_SIZE): Define variant for ELFv2.
	* src/powerpc/ffi.c (FLAG_ARG_NEEDS_PSAVE): Define.
	(discover_homogeneous_aggregate): New function.
	(ffi_prep_args64): Adjust start of param save area for ELFv2.
	Handle homogenous floating point struct parms.
	(ffi_prep_cif_machdep_core): Adjust space calculation for ELFv2.

src/arc/ffi.c
/* -----------------------------------------------------------------------
   ffi.c - Copyright (c) 2013  Synopsys, Inc. (www.synopsys.com)
   
   ARC Foreign Function Interface 

   Permission is hereby granted, free of charge, to any person obtaining
   a copy of this software and associated documentation files (the
   ``Software''), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED ``AS IS'', WITHOUT WARRANTY OF ANY KIND, EXPRESS
   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
   IN NO EVENT SHALL RENESAS TECHNOLOGY BE LIABLE FOR ANY CLAIM, DAMAGES OR
   OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
   OTHER DEALINGS IN THE SOFTWARE.
   ----------------------------------------------------------------------- */

#include <ffi.h>
#include <ffi_common.h>

#include <stdlib.h>
#include <stdint.h>

#include <sys/cachectl.h>

/* for little endian ARC, the code is in fact stored as mixed endian for
   performance reasons */
#if __BIG_ENDIAN__
#define CODE_ENDIAN(x) (x)
#else
#define CODE_ENDIAN(x) ( (((uint32_t) (x)) << 16) | (((uint32_t) (x)) >> 16))
#endif

/* ffi_prep_args is called by the assembly routine once stack
   space has been allocated for the function's arguments.  */

void
ffi_prep_args (char *stack, extended_cif * ecif)
{
  unsigned int i;
  void **p_argv;
  char *argp;
  ffi_type **p_arg;

  argp = stack;

  if (ecif->cif->rtype->type == FFI_TYPE_STRUCT)
    {
      *(void **) argp = ecif->rvalue;
      argp += 4;
    }

  p_argv = ecif->avalue;

  for (i = ecif->cif->nargs, p_arg = ecif->cif->arg_types;
       (i != 0); i--, p_arg++)
    {
      size_t z;
      int alignment;

      /* align alignment to 4 */
      alignment = (((*p_arg)->alignment - 1) | 3) + 1;

      /* Align if necessary.  */
      if ((alignment - 1) & (unsigned) argp)
	argp = (char *) ALIGN (argp, alignment);

      z = (*p_arg)->size;
      if (z < sizeof (int))
	{
	  z = sizeof (int);

	  switch ((*p_arg)->type)
	    {
	    case FFI_TYPE_SINT8:
	      *(signed int *) argp = (signed int) *(SINT8 *) (*p_argv);
	      break;

	    case FFI_TYPE_UINT8:
	      *(unsigned int *) argp = (unsigned int) *(UINT8 *) (*p_argv);
	      break;

	    case FFI_TYPE_SINT16:
	      *(signed int *) argp = (signed int) *(SINT16 *) (*p_argv);
	      break;

	    case FFI_TYPE_UINT16:
	      *(unsigned int *) argp = (unsigned int) *(UINT16 *) (*p_argv);
	      break;

	    case FFI_TYPE_STRUCT:
	      memcpy (argp, *p_argv, (*p_arg)->size);
	      break;

	    default:
	      FFI_ASSERT (0);
	    }
	}
      else if (z == sizeof (int))
	{
	  *(unsigned int *) argp = (unsigned int) *(UINT32 *) (*p_argv);
	}
      else
	{
	  if ((*p_arg)->type == FFI_TYPE_STRUCT)
	    {
	      memcpy (argp, *p_argv, z);
	    }
	  else
	    {
	      /* Double or long long 64bit.  */
	      memcpy (argp, *p_argv, z);
	    }
	}
      p_argv++;
      argp += z;
    }

  return;
}

/* Perform machine dependent cif processing.  */
ffi_status
ffi_prep_cif_machdep (ffi_cif * cif)
{
  /* Set the return type flag.  */
  switch (cif->rtype->type)
    {
    case FFI_TYPE_VOID:
      cif->flags = (unsigned) cif->rtype->type;
      break;

    case FFI_TYPE_STRUCT:
      cif->flags = (unsigned) cif->rtype->type;
      break;

    case FFI_TYPE_SINT64:
    case FFI_TYPE_UINT64:
    case FFI_TYPE_DOUBLE:
      cif->flags = FFI_TYPE_DOUBLE;
      break;

    case FFI_TYPE_FLOAT:
    default:
      cif->flags = FFI_TYPE_INT;
      break;
    }

  return FFI_OK;
}

extern void ffi_call_ARCompact (void (*)(char *, extended_cif *),
				extended_cif *, unsigned, unsigned,
				unsigned *, void (*fn) (void));

void
ffi_call (ffi_cif * cif, void (*fn) (void), void *rvalue, void **avalue)
{
  extended_cif ecif;

  ecif.cif = cif;
  ecif.avalue = avalue;

  /* If the return value is a struct and we don't have
     a return value address then we need to make one.  */
  if ((rvalue == NULL) && (cif->rtype->type == FFI_TYPE_STRUCT))
    {
      ecif.rvalue = alloca (cif->rtype->size);
    }
  else
    ecif.rvalue = rvalue;

  switch (cif->abi)
    {
    case FFI_ARCOMPACT:
      ffi_call_ARCompact (ffi_prep_args, &ecif, cif->bytes,
			  cif->flags, ecif.rvalue, fn);
      break;

    default:
      FFI_ASSERT (0);
      break;
    }
}

int
ffi_closure_inner_ARCompact (ffi_closure * closure, void *rvalue,
			     ffi_arg * args)
{
  void **arg_area, **p_argv;
  ffi_cif *cif = closure->cif;
  char *argp = (char *) args;
  ffi_type **p_argt;
  int i;

  arg_area = (void **) alloca (cif->nargs * sizeof (void *));

  /* handle hidden argument */
  if (cif->flags == FFI_TYPE_STRUCT)
    {
      rvalue = *(void **) argp;
      argp += 4;
    }

  p_argv = arg_area;

  for (i = 0, p_argt = cif->arg_types; i < cif->nargs;
       i++, p_argt++, p_argv++)
    {
      size_t z;
      int alignment;

      /* align alignment to 4 */
      alignment = (((*p_argt)->alignment - 1) | 3) + 1;

      /* Align if necessary.  */
      if ((alignment - 1) & (unsigned) argp)
	argp = (char *) ALIGN (argp, alignment);

      z = (*p_argt)->size;
      *p_argv = (void *) argp;
      argp += z;
    }

  (closure->fun) (cif, rvalue, arg_area, closure->user_data);

  return cif->flags;
}

extern void ffi_closure_ARCompact (void);

ffi_status
ffi_prep_closure_loc (ffi_closure * closure, ffi_cif * cif,
		      void (*fun) (ffi_cif *, void *, void **, void *),
		      void *user_data, void *codeloc)
{
  uint32_t *tramp = (uint32_t *) & (closure->tramp[0]);

  switch (cif->abi)
    {
    case FFI_ARCOMPACT:
      FFI_ASSERT (tramp == codeloc);
      tramp[0] = CODE_ENDIAN (0x200a1fc0);	/* mov r8, pcl  */
      tramp[1] = CODE_ENDIAN (0x20200f80);	/* j [long imm] */
      tramp[2] = CODE_ENDIAN (ffi_closure_ARCompact);
      break;

    default:
      return FFI_BAD_ABI;
    }

  closure->cif = cif;
  closure->fun = fun;
  closure->user_data = user_data;
  cacheflush (codeloc, FFI_TRAMPOLINE_SIZE, BCACHE);

  return FFI_OK;
}
